dot/state/epoch.go

```diff
@@ -440,50 +452,105 @@ func (s *EpochState) GetLatestConfigData() (*types.ConfigData, error) {
+		slotDuration, err := s.GetSlotDuration()
+		if err != nil {
+			return 0, fmt.Errorf("getting slot duration: %w", err)
+		}
+
+		return uint64(time.Now().UnixNano()) / uint64(slotDuration.Nanoseconds()), nil
 	}
 
-	return s.epochLength*epoch + firstSlot, nil
+	if errors.Is(err, ErrNoFirstNonOriginBlock) && epoch != 0 {
+		return 0, fmt.Errorf(
+			"%w: first non origin block is needed for epoch %d",
+			ErrNoFirstNonOriginBlock,
+			epoch)
+	}
+	return s.epochLength*epoch + chainFirstSlotNumber, nil
 }
 
-// GetEpochFromTime returns the epoch for a given time
-func (s *EpochState) GetEpochFromTime(t time.Time) (uint64, error) {
-	slotDuration, err := s.GetSlotDuration()
+// retrieveFirstNonOriginBlockSlot returns the slot number of the very first non origin block
+// if there is more than one first non origin block then it uses the block hash to check ancestry
+// e.g to return the correct slot number for a specific fork
+func (s *EpochState) retrieveFirstNonOriginBlockSlot(blockHash common.Hash) (uint64, error) {
+	firstNonOriginHashes, err := s.blockState.GetHashesByNumber(1)
 	if err != nil {
-		return 0, err
+		return 0, fmt.Errorf("getting hashes using number 1: %w", err)
 	}
 
-	firstSlot, err := s.baseState.loadFirstSlot()
-	if err != nil {
-		return 0, err
+	if len(firstNonOriginHashes) == 0 {
+		return 0, ErrNoFirstNonOriginBlock
+	}
+
+	var firstNonOriginBlockHash common.Hash
+	if len(firstNonOriginHashes) == 1 {
+		firstNonOriginBlockHash = firstNonOriginHashes[0]
+	} else {
+		blockHeader, err := s.blockState.GetHeader(blockHash)
+		if err != nil {
+			return 0, fmt.Errorf("getting block by header: %w", err)
+		}
+
+		if blockHeader.Number == 1 {
+			return blockHeader.SlotNumber()
+		}
+
+		for _, hash := range firstNonOriginHashes {
+			isDescendant, err := s.blockState.IsDescendantOf(hash, blockHash)
+			if err != nil {
+				return 0, fmt.Errorf("while checking ancestry: %w", err)
+			}
+
+			if isDescendant {
+				firstNonOriginBlockHash = hash
+				break
+			}
+		}
 	}
 
-	slot := uint64(t.UnixNano()) / uint64(slotDuration.Nanoseconds())
+	firstNonGenesisHeader, err := s.blockState.GetHeader(firstNonOriginBlockHash)
+	if err != nil {
+		return 0, fmt.Errorf("getting first non genesis block by hash: %w", err)
+	}
 
-	if slot < firstSlot {
-		return 0, errors.New("given time is before network start")
+	chainFirstSlotNumber, err := firstNonGenesisHeader.SlotNumber()
+	if err != nil {
+		return 0, fmt.Errorf("getting slot number: %w", err)
 	}
 
-	return (slot - firstSlot) / s.epochLength, nil
+	return chainFirstSlotNumber, nil
 }
 
-// SetFirstSlot sets the first slot number of the network
-func (s *EpochState) SetFirstSlot(slot uint64) error {
-	// check if block 1 was finalised already; if it has, don't set first slot again
-	header, err := s.blockState.GetHighestFinalisedHeader()
+// GetEpochFromTime returns the epoch for a given time
+func (s *EpochState) GetEpochFromTime(t time.Time, blockHash common.Hash) (uint64, error) {
```

Comment:
Removed the unneeded method `GetEpochFromTime`, the explanation regards the block hash is because we need to be fork aware when retrieving the first slot of the chain given that we might have many blocks `#1` and we should retrieve the correct one given the ancestry of the `blockHash`, this explanation is at `retrieveFirstNonOriginBlockSlot` method